src/Evolution/Systems/CurvedScalarWave/Worldtube/SingletonActions/ChangeSlabSize.hpp

```diff
@@ -0,0 +1,90 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <cstddef>
+#include <memory>
+
+#include "DataStructures/DataBox/DataBox.hpp"
+#include "DataStructures/DataBox/Prefixes.hpp"
+#include "Evolution/Systems/CurvedScalarWave/Worldtube/Inboxes.hpp"
+#include "Evolution/Systems/CurvedScalarWave/Worldtube/Tags.hpp"
+#include "Parallel/AlgorithmExecution.hpp"
+#include "Parallel/GlobalCache.hpp"
+#include "ParallelAlgorithms/EventsAndTriggers/Tags.hpp"
+#include "Time/Actions/ChangeSlabSize.hpp"
+#include "Time/Tags.hpp"
+#include "Time/TimeStepId.hpp"
+#include "Time/TimeSteppers/TimeStepper.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/TMPL.hpp"
+
+namespace CurvedScalarWave::Worldtube::Actions {
+
+/*!
+ * \brief Waits for the data from all neighboring elements and changes the slab
+ * size if a change in the global time step is detected.
+ * \details We check the slab size of the time step id sent by the elements. If
+ * this is different from the slab size currently used by the worldtube
+ * singleton, we assume a global slab size change has occurred in the elements
+ * and adjust the worldtube slab size accordingly.
+ */
+struct ChangeSlabSize {
+  static constexpr size_t Dim = 3;
+  using inbox_tags = tmpl::list<
+      ::CurvedScalarWave::Worldtube::Tags::SphericalHarmonicsInbox<Dim>>;
+
+  template <typename DbTagsList, typename... InboxTags, typename Metavariables,
+            typename ArrayIndex, typename ActionList,
+            typename ParallelComponent>
+  static Parallel::iterable_action_return_t apply(
+      db::DataBox<DbTagsList>& box, tuples::TaggedTuple<InboxTags...>& inboxes,
+      const Parallel::GlobalCache<Metavariables>& /*cache*/,
+      const ArrayIndex& /*array_index*/, ActionList /*meta*/,
+      const ParallelComponent* const /*meta*/) {
+    const auto& time_step_id = db::get<::Tags::TimeStepId>(box);
+    auto& inbox = tuples::get<Tags::SphericalHarmonicsInbox<Dim>>(inboxes);
+    if (inbox.empty()) {
+      return {Parallel::AlgorithmExecution::Retry, std::nullopt};
+    }
+    ASSERT(inbox.size() == 1,
+           "Received data from two different time step ids.");
```

Comment:
Does your communication pattern couple things tightly enough that it shouldn't be possible to receive two messages without going through the action loop in between?  For example, in (GTS) DG your neighbor can send boundary data twice between your steps.

Reply:
yes, at the moment it's coupled this tightly.